diff -urN original/configure patched/configure
--- original/configure	2010-05-07 23:10:16.000000000 -0400
+++ patched/configure	2010-07-06 17:07:02.000000000 -0400
@@ -2731,6 +2731,7 @@
     case "${canonical}" in
       i[3456]86-* )  machine=intel386 ;;
       powerpc-* )    machine=macppc ;;
+      arm-* )        machine=arm ;;
       x86_64-* )     machine=amdx86-64 ;;
       * )            unported=yes ;;
     esac
diff -urN original/src/unexmacosx.c patched/src/unexmacosx.c
--- original/src/unexmacosx.c	2010-04-03 18:26:11.000000000 -0400
+++ patched/src/unexmacosx.c	2010-07-06 17:13:13.000000000 -0400
@@ -955,6 +955,11 @@
 	    /* nothing to do for prebound lazy pointer */
 	    break;
 #endif
+
+          case GENERIC_RELOC_LOCAL_SECTDIFF:
+            /* no idea what this is, but we ignore it anyway */
+            break;
+
 	  default:
 	    unexec_error ("unrelocate: %s:%d cannot handle scattered type = %d",
 			  name, i, sc_reloc_info->r_type);
