--- darwin.h~	2010-04-11 01:09:13.000000000 +1000
+++ darwin.h	2011-09-08 17:29:34.000000000 +1000
@@ -220,8 +220,11 @@
 #define NO_ABORT
 
 /* Do not define matherr in floatfns.c.  */
+#undef NO_MATHERR
 #define NO_MATHERR
 
+
+
 /* The following solves the problem that Emacs hangs when evaluating
    (make-comint "test0" "/nodir/nofile" nil "") when /nodir/nofile
    does not exist.  Also, setsid is not allowed in the vfork child's
